### ISSUE TITLE
Issues with Implementing Morp Relationship with Product and Variant Models - getAttribute() on null

### ISSUE THREAD

**Issue by Vinay-Indoria** [OFFLINE-GmbH/oc-mall-plugin#349]
Hi,
So here is the use-case, I need to attach the product to a Custom Model suppose the **Bucket** model but my store has Product as well Product with Variant.
Since we have 2 different model for Product and Variant in the plugin I need to handle connection to both the models(*please suggest if there is another way to achieve this*). The structure for the database looks something like below:
```
products
    id - integer
    name - string

variants
    id - integer
    product_id - integer
    name - string

buckets
    id - integer
    name - string

bucket_product_or_variants
    bucket_id - integer
    bucket_product_or_variant_id - integer
    bucket_product_or_variant_type - string
```
The code in the Custom Bucket model is as follow
```PHP
public $morphedByMany = [
        'products'  => ['Offline\Mall\Models\Product', 'name' => 'bucket_product_or_variant'],
        'variants' => ['Offline\Mall\Models\Variant', 'name' => 'bucket_product_or_variant']
    ];
```
In my **plugins.php** I tried to extend both product and variant models 
```PHP
use Offline\Mall\Models\Product as ProductModel;
use Offline\Mall\Models\Variant as VariantModel;

    public function boot(){
        
        ProductModel::extend(function ($model){
            $model->morphToMany['buckets'] = ['Author\Buckets\Models\Bucket', 'name' => 'bucket_product_or_variant'];
        });

        VariantModel::extend(function ($model){
            $model->morphToMany['buckets'] = ['Author\Buckets\Models\Bucket', 'name' => 'bucket_product_or_variant'];
        });
    }
```
All this is fine and I have data inside my intermediate table(*bucket_product_or_variants*) with id's of the relevant models but when I try to access the edit screen of the added bucket record I receive an exception - **call to function getAttribute() on null** which I am not able to figure out the cause for!

![Exception](https://user-images.githubusercontent.com/1214829/67196950-69544180-f419-11e9-86ca-22fcf8c7cac2.png)

Here upon basic debugging on **line 227** for the Variant Model, `$this->product` is empty.

Please suggest if I missed out on anything!

**Comment by tobias-kuendig** [comment]
It looks like your Variant model does not belong to a Product model, which it must.

**Comment by Vinay-Indoria** [comment]
belongs to for Variant Model @tobias-kuendig 
```PHP
public $belongsTo = [
        'product'      => Product::class,
        'cart_product' => CartProduct::class,
        'image_sets'   => [ImageSet::class, 'key' => 'image_set_id'],
    ];
```

**Comment by tobias-kuendig** [comment]
The Variant model you have loaded in the screenshot above, does it have a value set in the `product_id` db column? If `$this->product` is null, this means that there is no Product model associated.

**Comment by Vinay-Indoria** [comment]
Yes the entry for the variant in `offline_mall_product_variants` table has the value set for product_id column.
But when we refer the table which is storing the polymorphic relation it looks like follow:

bucket_id | bucket_product_or_variant_id | bucket_product_or_variant_type
-----------|----------------------------------|-----------------------------------
1 | 1 | mall.product
1 | 2 | mall.variant

This table should also have a product_id against the variant record? but if that is the case the entire structure goes for toss and it won't make sense!

**Comment by tobias-kuendig** [comment]
Just debug the line where the exception is thrown. What is the value uf `$this` and `$this->product_id` in the context of the error from your first comment?

**Comment by Vinay-Indoria** [comment]
`$this->product_id` is `null`
Here is the output for `$this`:

```PHP
Variant {#628 ▼
  +slugs: []
  +nullable: array:1 [▼
    0 => "image_set_id"
  ]
  +table: "offline_mall_product_variants"
  +dates: array:1 [▶]
  +with: array:4 [▼
    0 => "product.additional_prices"
    1 => "image_sets"
    2 => "prices"
    3 => "additional_prices"
  ]
  +implement: array:1 [▼
    0 => "@RainLab.Translate.Behaviors.TranslatableModel"
  ]
  +translatable: array:3 [▶]
  +casts: array:6 [▶]
  +rules: array:6 [▼
    "name" => "required"
    "product_id" => "required"
    "stock" => "required|integer"
    "weight" => "nullable|integer"
    "published" => "boolean"
    "allow_out_of_stock_purchases" => "boolean"
  ]
  +attachMany: array:2 [▼
    "temp_images" => "System\Models\File"
    "downloads" => "System\Models\File"
  ]
  +belongsTo: array:3 [▼
    "product" => "OFFLINE\Mall\Models\Product"
    "cart_product" => "OFFLINE\Mall\Models\CartProduct"
    "image_sets" => array:2 [▶]
  ]
  +hasMany: array:5 [▶]
  +morphMany: array:2 [▼
    "customer_group_prices" => array:2 [▶]
    "additional_prices" => array:2 [▶]
  ]
  #fillable: array:10 [▼
    0 => "product_id"
    1 => "user_defined_id"
    2 => "image_set_id"
    3 => "stock"
    4 => "name"
    5 => "published"
    6 => "weight"
    7 => "allow_out_of_stock_purchases"
    8 => "mpn"
    9 => "gtin"
  ]
  +attributes: array:1 [▼
    "id" => 2
  ]
  #jsonable: []
  +duplicateCache: true
  #connection: "mysql"
  #primaryKey: "id"
  #keyType: "int"
  +incrementing: true
  #withCount: []
  #perPage: 15
  +exists: true
  +wasRecentlyCreated: false
  #original: array:1 [▼
    "id" => 2
  ]
  #changes: []
  #dateFormat: null
  #appends: []
  #dispatchesEvents: []
  #observables: []
  #relations: []
  #touches: []
  +timestamps: true
  #hidden: []
  #visible: []
  #guarded: array:1 [▼
    0 => "*"
  ]
  +hasOne: []
  +belongsToMany: []
  +morphTo: []
  +morphOne: []
  +morphToMany: array:1 [▼
    "buckets" => array:2 [▼
      0 => "Author\Buckets\Models\Bucket"
      "name" => "bucket_product_or_variant"
    ]
  ]
  +morphedByMany: []
  +attachOne: []
  +hasManyThrough: []
  #emitterSingleEventCollection: []
  #emitterEventCollection: array:2 [▼
    "model.saveInternal" => array:1 [▶]
    "model.beforeSave" => array:1 [▶]
  ]
  #emitterEventSorted: []
  #extensionData: array:4 [▼
    "extensions" => []
    "methods" => []
    "dynamicMethods" => []
    "dynamicProperties" => []
  ]
  +sessionKey: null
  #validationErrors: null
  #validationDefaultAttrNames: []
  #forceDeleting: false
  #money: DefaultMoney {#293 ▼
    #twig: Environment {#294 ▼
      -charset: "UTF-8"
      -loader: Loader {#295 ▶}
      -debug: false
      -autoReload: true
      -cache: NullCache {#296}
      -lexer: null
      -parser: null
      -compiler: null
      -baseTemplateClass: "\Twig\Template"
      -globals: []
      -resolvedGlobals: null
      -loadedTemplates: null
      -strictVariables: false
      -templateClassPrefix: "__TwigTemplate_"
      -originalCache: false
      -extensionSet: ExtensionSet {#297 ▶}
      -runtimeLoaders: []
      -runtimes: []
      -optionsHash: "["Twig\\Extension\\CoreExtension","Twig\\Extension\\EscaperExtension","Twig\\Extension\\OptimizerExtension","System\\Twig\\Extension"]:7:2:2.11.3:0:\Twig\Templa ▶"
    }
  }
  #defaultCurrency: null
}
```

**Comment by tobias-kuendig** [comment]
From the error screenshot of your former comment, can you show me the full stack trace?

**Comment by Vinay-Indoria** [comment]
```
#	CALLED CODE	DOCUMENT	LINE
66	OFFLINE\Mall\Models\Variant->getAttribute(…)	~/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php	1382
65	Illuminate\Database\Eloquent\Model->__get(…)	~/vendor/october/rain/src/Extension/ExtendableTrait.php	360
64	October\Rain\Database\Model->extendableGet(…)	~/vendor/october/rain/src/Database/Model.php	617
63	October\Rain\Database\Model->__get(…)	~/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php	689
62	Illuminate\Database\Eloquent\Builder->Illuminate\Database\Eloquent\{closure}(…)		
61	array_map(…)	~/vendor/laravel/framework/src/Illuminate/Support/Collection.php	861
60	Illuminate\Support\Collection->map(…)	~/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php	690
59	Illuminate\Database\Eloquent\Builder->pluck(…)	~/vendor/october/rain/src/Database/Relations/BelongsToMany.php	378
58	October\Rain\Database\Relations\BelongsToMany->allRelatedIds(…)	~/vendor/october/rain/src/Database/Relations/BelongsToMany.php	358
57	October\Rain\Database\Relations\BelongsToMany->getSimpleValue()	~/vendor/october/rain/src/Database/Concerns/HasRelationships.php	683
56	October\Rain\Database\Model->getRelationValue(…)	~/modules/backend/classes/FormField.php	693
55	Backend\Classes\FormField->getFieldNameFromData(…)	~/modules/backend/classes/FormField.php	623
54	Backend\Classes\FormField->getValueFromData(…)	~/modules/backend/widgets/Form.php	1084
53	Backend\Widgets\Form->getFieldValue(…)	~/modules/backend/widgets/Form.php	843
52	Backend\Widgets\Form->makeFormField(…)	~/modules/backend/widgets/Form.php	688
51	Backend\Widgets\Form->addFields(…)	~/modules/backend/widgets/Form.php	534
50	Backend\Widgets\Form->defineFormFields()	~/modules/backend/widgets/Form.php	153
49	Backend\Widgets\Form->bindToController()	~/modules/backend/behaviors/FormController.php	175
48	Backend\Behaviors\FormController->initForm(…)	~/modules/backend/behaviors/FormController.php	290
47	Backend\Behaviors\FormController->update(…)		
46	call_user_func_array(…)	~/vendor/october/rain/src/Extension/ExtendableTrait.php	409
45	Backend\Classes\Controller->extendableCall(…)	~/modules/backend/classes/Controller.php	192
44	Backend\Classes\Controller->__call(…)		
43	call_user_func_array(…)	~/modules/backend/classes/Controller.php	397
42	Backend\Classes\Controller->execPageAction(…)	~/modules/backend/classes/Controller.php	285
41	Backend\Classes\Controller->run(…)	~/modules/backend/classes/BackendController.php	161
40	Backend\Classes\BackendController->run(…)		
39	call_user_func_array(…)	~/vendor/laravel/framework/src/Illuminate/Routing/Controller.php	54
38	Illuminate\Routing\Controller->callAction(…)	~/vendor/laravel/framework/src/Illuminate/Routing/ControllerDispatcher.php	45
37	Illuminate\Routing\ControllerDispatcher->dispatch(…)	~/vendor/laravel/framework/src/Illuminate/Routing/Route.php	212
36	Illuminate\Routing\Route->runController()	~/vendor/laravel/framework/src/Illuminate/Routing/Route.php	169
35	Illuminate\Routing\Route->run()	~/vendor/laravel/framework/src/Illuminate/Routing/Router.php	658
34	Illuminate\Routing\Router->Illuminate\Routing\{closure}(…)	~/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php	30
33	Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(…)	~/modules/backend/classes/BackendController.php	68
32	Backend\Classes\BackendController->Backend\Classes\{closure}(…)	~/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php	131
31	Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(…)	~/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php	53
30	Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(…)	~/vendor/laravel/framework/src/Illuminate/Routing/Middleware/SubstituteBindings.php	41
29	Illuminate\Routing\Middleware\SubstituteBindings->handle(…)	~/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php	149
28	Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(…)	~/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php	53
27	Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(…)	~/vendor/laravel/framework/src/Illuminate/View/Middleware/ShareErrorsFromSession.php	49
26	Illuminate\View\Middleware\ShareErrorsFromSession->handle(…)	~/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php	149
25	Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(…)	~/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php	53
24	Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(…)	~/vendor/laravel/framework/src/Illuminate/Session/Middleware/StartSession.php	63
23	Illuminate\Session\Middleware\StartSession->handle(…)	~/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php	149
22	Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(…)	~/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php	53
21	Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(…)	~/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/AddQueuedCookiesToResponse.php	37
20	Illuminate\Cookie\Middleware\AddQueuedCookiesToResponse->handle(…)	~/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php	149
19	Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(…)	~/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php	53
18	Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(…)	~/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/EncryptCookies.php	66
17	Illuminate\Cookie\Middleware\EncryptCookies->handle(…)	~/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php	149
16	Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(…)	~/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php	53
15	Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(…)	~/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php	102
14	Illuminate\Pipeline\Pipeline->then(…)	~/vendor/laravel/framework/src/Illuminate/Routing/Router.php	660
13	Illuminate\Routing\Router->runRouteWithinStack(…)	~/vendor/laravel/framework/src/Illuminate/Routing/Router.php	635
12	Illuminate\Routing\Router->runRoute(…)	~/vendor/laravel/framework/src/Illuminate/Routing/Router.php	601
11	Illuminate\Routing\Router->dispatchToRoute(…)	~/vendor/october/rain/src/Router/CoreRouter.php	20
10	October\Rain\Router\CoreRouter->dispatch(…)	~/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php	176
9	Illuminate\Foundation\Http\Kernel->Illuminate\Foundation\Http\{closure}(…)	~/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php	30
8	Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(…)	~/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/CheckForMaintenanceMode.php	46
7	Illuminate\Foundation\Http\Middleware\CheckForMaintenanceMode->handle(…)	~/vendor/october/rain/src/Foundation/Http/Middleware/CheckForMaintenanceMode.php	24
6	October\Rain\Foundation\Http\Middleware\CheckForMaintenanceMode->handle(…)	~/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php	149
5	Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(…)	~/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php	53
4	Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(…)	~/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php	102
3	Illuminate\Pipeline\Pipeline->then(…)	~/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php	151
2	Illuminate\Foundation\Http\Kernel->sendRequestThroughRouter(…)	~/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php	116
1	Illuminate\Foundation\Http\Kernel->handle(…)	~/index.php	43
```
Screenshots if the above is not clear, tried different ways to get the stack trace come out properly here but only the above was readable.

![Exception-StackTace1](https://user-images.githubusercontent.com/1214829/67273393-f790fc00-f4db-11e9-9eb6-82e4fef50322.png)

![Exception-StackTace2](https://user-images.githubusercontent.com/1214829/67273413-ff50a080-f4db-11e9-8a7d-08d6c8c1de64.png)

![Exception-StackTrace3](https://user-images.githubusercontent.com/1214829/67273422-04adeb00-f4dc-11e9-8a0b-d4705768c75f.png)

**Comment by tobias-kuendig** [comment]
Can you replace [this line](https://github.com/OFFLINE-GmbH/oc-mall-plugin/blob/80a210ab8645a8668b70212488dfec91832cbcce/models/Variant.php#L225) with the code below and see if it fixes the issue for you?

```php
if (!$this->product) {
    return $originalValue;
}
$parentValues = $this->product->getAttribute($attribute);
```

**Comment by Vinay-Indoria** [comment]
yus this solves the issue!! Thank you..

**Comment by tobias-kuendig** [comment]
This is fixed in 18b0079. Thanks!
